test/libs/helper-functions.js

```diff
@@ -0,0 +1,194 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* eslint-env browser */
+
+// Each service worker that is registered should be given a unique
+// scope. To achieve this we register it with a scope the same as
+// an iframe's src that is unique for each test.
+// Service workers will then be made to claim pages on this scope -
+// i.e. the iframe
+var getIframe = function() {
+  return new Promise(resolve => {
+    var existingIframe = document.querySelector('.js-test-iframe');
+    if (existingIframe) {
+      return resolve(existingIframe);
+    }
+
+    var newIframe = document.createElement('iframe');
+    newIframe.classList.add('js-test-iframe');
+    newIframe.src = '/test/iframe/' + Math.random();
+    newIframe.addEventListener('load', () => {
+      resolve(newIframe);
+    });
+    document.body.appendChild(newIframe);
+  });
+};
+
+window.testHelper = {
+  unregisterAllRegistrations: function() {
+    return navigator.serviceWorker.getRegistrations()
+      .then(registrations => {
+        if (registrations.length === 0) {
+          return Promise.resolve();
+        }
+
+        var unregisterPromises = [];
+        for (var i = 0; i < registrations.length; i++) {
+          unregisterPromises.push(
+            registrations[i].unregister()
+          );
+        }
+        return Promise.all(unregisterPromises);
+      });
+  },
+
+  clearAllCaches: function() {
+    return window.caches.keys()
+      .then(cacheNames => {
+        if (cacheNames.length === 0) {
+          return Promise.resolve();
+        }
+
+        var cacheTaskPromises = [];
+        for (var i = 0; i < cacheNames.length; i++) {
+          cacheTaskPromises.push(window.caches.delete(cacheNames[i]));
+        }
+        return Promise.all(cacheTaskPromises);
+      });
+  },
+
+  // Waiting for a service worker to install is handy if you only care
+  // about testing events that have occured in the install event
+  installSW: function(swUrl) {
+    return new Promise((resolve, reject) => {
+      var iframe;
+      getIframe()
+      .then(newIframe => {
+        var options = null;
+        if (newIframe) {
+          options = {scope: iframe.contentWindow.location.pathname};
+          iframe = newIframe;
+        }
+
+        return navigator.serviceWorker.register(swUrl, options);
+      })
+      .then(registration => {
+        if (registration.installing === null) {
+          throw new Error(swUrl + ' already installed.');
+        }
+
+        // We unregister all service workers after each test - this should
+        // always trigger an install state change
+        registration.installing.onstatechange = function() {
+          if (this.state !== 'installed') {
+            return;
+          }
+
+          resolve(iframe);
+        };
+      })
+      .catch(err => {
```

Comment:
It was reproducible when we found it. Don't think we documented it
anywhere.

On Tue, 12 Jan 2016, 18:38 Jeffrey Posnick <notifications@github.com> wrote:

> In test/libs/helper-functions.js
> <https://github.com/GoogleChrome/sw-toolbox/pull/70#discussion_r49494355>:
>
> > +      .then(registration => {
> > +        if (registration.installing === null) {
> > +          throw new Error(swUrl + ' already installed.');
> > +        }
> > +
> > +        // We unregister all service workers after each test - this should
> > +        // always trigger an install state change
> > +        registration.installing.onstatechange = function() {
> > +          if (this.state !== 'installed') {
> > +            return;
> > +          }
> > +
> > +          resolve(iframe);
> > +        };
> > +      })
> > +      .catch(err => {
>
> Was that behavior an expected edge case, or a bug? Can you reproduce it?
>
> Alternatively, you can just return a promise chain that starts with
> getIframe() instead of constructing a new Promise wrapper.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleChrome/sw-toolbox/pull/70/files#r49494355>.
>